
August 2015  
Modified September 2017, October 2018

Author: FBB

** reproduce the result in **
http://www.mdrc.org/sites/default/files/What%20Strategies%20Work%20for%20the%20Hard%20FR.pdf 


# This is Yunhe Cui's HW6-Assignment 3

In [1]:
import os
import sys
import numpy as np
import pylab as pl

%pylab inline

#from IPython.display import Image
#Image(filename='../plotsforclasses/NYCReentryprogram_title.png')



Populating the interactive namespace from numpy and matplotlib


In [2]:
#Image(filename='../plotsforclasses/NYCReentryprogram.png')

# NULL HYPOTHESIS: the % of former prisoners employed after release is the same or lower for candidates who participated in the program as for the control group, significance level p=0.05

** $H_0: P_0 - P_1 \geq$    0 **
    
** $H_a: P_0 - P_1 $< 0    **
    
    
** $\alpha$ = 0.05    **

** this is a TEST OF PROPORTIONS. we use the Binomial distribution since it is a yes/no (bernulli) test for each subject: the former inmate was or was not ever employed in a CEO transitional job (second row in the table above): **

** $P_0=0.035, P_1=0.701$**

In [3]:

# first thing: choose your significance threshold
alpha=0.05



# we like fractions better then percentages. as a rule of thumb, either use fractions or counts. 
# 0 for control group, 1 for program group

P_0 = 3.5 * 0.01 
P_1 = 70.1 * 0.01

if P_0 - P_1 >= 0:
    # we are done
    print ("the Null holds")
else:
    print ("we must assess the statistical significance")

#sample sizes 

n_0 = 409
n_1 = 564
    
#lets get the counts by multiplying by the sample size

Nt_0 = P_0 * n_0
Nt_1 = P_1 * n_1


we must assess the statistical significance


# WHICH  TEST?

samples, categorical data
 
 ** TWO OPTIONS z test, or chi-square test.  **

# START WITH Z TEST

** the z test compares the stanrard deviation of the expected distribution and the observed result. it tells you literally how many standard deviations from the tail an observation is, under the _assumption of normality_ **


In [4]:
#Image("ztest.png")
#https://onlinecourses.science.psu.edu/stat414/node/268

must define the sample standard deviation (standard error)

In [5]:
#define the sample proportion first
sp = (P_0 * n_0 + P_1 * n_1) / (n_1 + n_0)
print (sp)

0.421047276465


** standard deviation of the sampling distribution the distribution is Binomial, the binomial stdev is **

(see a proof here!: http://stats.stackexchange.com/questions/29641/standard-error-for-the-mean-of-a-sample-of-binomial-random-variables!): 

$\sqrt{\frac{p(1 - p)}{n}}$

for 2 samples this becomes 

$\sqrt{ \frac{ \hat{p}(1 - \hat{p})} {n0} + \frac{ \hat{p}(1 - \hat{p})} {n1} }$

cfr: page 138 of Statistics in a Nutshell, eq. 5.12 and here http://stattrek.com/hypothesis-test/difference-in-proportions.aspx?Tutorial=AP

** $\hat{p} =\frac{p_0  n_0 + p_1  n_1}{n_0+n_1}$ **


Note that in the online version of Statistics in a Nutshell, $\bar{x}$  should be  $\hat{p}$!!

In [6]:
# I am goonna create a little one line function to calculate the standard error
# and to calculate p

p = lambda p0, p1, n0, n1: (p0 * n0 + p1 * n1) / (n0 + n1)
#standard error
se = lambda p, n0, n1: np.sqrt(p * (1 - p) * (1.0 / n0 + 1.0 / n1)) 

** z score: how many standard deviation away from the population parameter is my statistic? **

 $z=\frac{P_1-P_0}{\sigma}$

In [7]:
zscore = lambda p0, p1, s : (p0 - p1) / s
z_2y = zscore(P_1, P_0, se(p(P_0, P_1, n_0, n_1), n_0, n_1))
print (z_2y)


20.7697865408


note that using p0-p1 or p1-p0 at the numerator is equivalent because the standardizes normal value of z has mean 0 (see image below) so that we can use the absolute value of the z score, or equivalently look for $P[Z<z]$ if z is positive, and $P[Z>z]$ if z is negative.

In [8]:
#Image('http://intersci.ss.uci.edu/wiki/images/3/3a/Normal01.jpg')

* if $p<\alpha$ : reject H0 **

IMPORTANT!! note that this P in the bottom line of the table is not the p-value, but 

p-value = 1-P

In [9]:
## p-value for employment after 2 years: 
## since the largest number we read off the table for is (way) smaller 
## than the value for our statistic 
## our p-value will be smaller than it would be if calculated using 
## (e.g.) .9998 (and in fact using 1.0000 which is the largest number 
## in the table). Using 0.9998 is a **conservative** approach. 

p_2y = 1 - 0.9984


def report_result(p,a):
    print ('is the p value ' + 
           '{0:.2f} smaller than the critical value {1:.2f}?'.format(p,a))
    if p < a:
        print ("YES!")
    else: 
        print ("NO!")
    
    print ('the Null hypothesis is {}'.format(\
                            'rejected' if p < a  else 'not rejected') )

    
report_result(p_2y, alpha)

is the p value 0.00 smaller than the critical value 0.05?
YES!
the Null hypothesis is rejected


# 1


# what if we used the values for where the former inmate was or was not "Convicted of a felony" (row 10) in the Recidivism (Years 1-3)?

## NULL HYPOTHESIS: the % of former prisoners who was convicted of a felony employed in the Recidivism after release is the same or higher for candidates who participated in the program as for the control group, significance level p=0.05


** $H_0: P_0 - P_1 $<=   0 **
    
** $H_a: P_0 - P_1 $> 0    **
    
    
** $\alpha$ = 0.05    **

 $P_0 = 11.7*0.01 = 0.117$   
 $P_1= 10.0 * 0.01 = 0.1$

 look up data table and insert the appropriate values to get the appropriate result! you can use the functions I defined above, with different arguments. 

In [18]:
P_0 = 0.117
P_1 = 0.1
n_0 = 409
n_1 = 568

if P_0 - P_1 <= 0:
    # we are done
    print ("the Null holds")
else:
    print ("we must assess the statistical significance")

# conduct the z test
z_3y = zscore(P_1, P_0, se(p(P_0, P_1, n_0, n_1), n_0, n_1))
print ("z score is %f"%z_3y)

we must assess the statistical significance
z score is -0.847640


In [19]:
# We could look up the z-score chart and find the P[Z>z] is 0.8023
p_3y = 1 - 0.8023
print (p_3y)

0.1977


In [20]:
report_result(p_3y, alpha)

is the p value 0.20 smaller than the critical value 0.05?
NO!
the Null hypothesis is not rejected


# Now lets do it with the $\chi^2$ test

this analysis can also be done with the $\chi^2$ test, and the  $\chi^2$ distribution, 

see  flowcharts for how to choose a test and "Statistics In a Nutshell Chapter 4", or http://math.hws.edu/javamath/ryan/ChiSquare.html (if you are really just interested in the formula at face value)

** The chisq statistics tests the statistics calculated as : **

$\chi^2 = \sum_{i} \frac{(observation_i - expectation_i)^2}{expectation_i}$

 against a chi sq distribution.
 If we talk about sample fractions  that is 

$\chi^2 = \sum_i \frac{(f_{i,observed} - f_{i,expectated})^2}{f_{i,expected}}$

 Where _i_ indicates the sum over _each cell_.
 turns out this quantity is distributed according to a chi square distribution, so if i get the $\chi^2$ statistics i can compare it to the full chisq distribution and see how far in the tail it is

 The trickiest part (but not that tricky) is to figure out how to construct the table of values. please see Statistics In a Nutshell Chapter 4, for our data for example: Thisis called a CONTINGENCY TABLE

|                 |     success         | failure|    |               
|-----------------|:-------------------:|:-------------------:|---------------------------|
| test sample     | number of successes in test    | number of failures in test    | number members of test sample |
| control sample  | number of successes in control | number of failures in control | number members of control sample| 
|                 | total successes                |  total failures               | number of all members         |

|employed in subsadized job |     employed          | not employed     |                   
|---------------------------|:---------------------:|------------------|---------------------------|
| test sample               | $0.701*564$           | $0.299*564$      | 564                       |
| control sample            | $0.035*409$           | $0.965*4.09$     | 409                       |
|                           |                       |                  |                           |
| total                     | 409.679               |  562.912         | 973                       |

 for each cell the expected ratio is the product of the total of all rows and all columns, devided by the total

** FOR EACH CELL: **
**  expected_cell = $\frac{row~tot * col~tot}{total}$ **


In [22]:
#from evalChisq import evalChisq

In [21]:

# Ntot = 973 # a + b + c + d = tot

# sample_values = np.array([[0.701 * 564, 0.299 * 564], [0.0305 * 409, 0.965 * 409]])

# print (evalChisq(sample_values))

This number must be compared to the chi sq distribution. 
You must calculate the number of degrees of freedom forthis experiment. 
Generally: DOF = Number of observations - number of Independent Variables

so here DOF = 1. Now you can look at the table below and draw conclusions about the rejection of the Null

In [23]:
#Image("http://passel.unl.edu/Image/Namuth-CovertDeana956176274/chi-sqaure%20distribution%20table.PNG")

432 is hella larger then 3.84

why am i mentioning 3.84?

how does the chi square statistic that you derived compare?

please  state what that means in terms of your Null hypothesis in a markdown cell below!

### We can see the Chi-square distribution, for p = 0.05 and DOF = 1, we could get 3.84 for the Chi-square critical value. The p-value we got in this problem is 432 which is much higher than 3.84. So that we could confidently reject the Null hypothesis which is ever employed in a CEO transitional job has no influence on the employment.  

# 2
lets see what the chi sq statistics says about the conviction for fellonies (row 10)

In [30]:
P_0 = 0.117
P_1 = 0.1
n_0 = 409
n_1 = 568
obeservation1 = n_0 * P_0
lr = n_0 * (1 - P_0)
ul = n_1 * P_1
ur = n_1 * (1 - P_1)
print (ll, lr, ul, ur)
print (ll + ul)
print (lr + ur)

(47.853, 361.147, 56.800000000000004, 511.2)
104.653
872.347


DOF = 1 Thus we can see the Chi-square distribution, for p = 0.05 and DOF = 1, we could get 3.84 for the propbability of a larger value of x**2

$\chi^2 = \sum_{i} \frac{(observation_i - expectation_i)^2}{expectation_i}$



In [33]:
exp12 = (568 * 0.1 + 409 * 0.117) / 2
exp34 = (568 * 0.9 + 409 * 0.883) / 2
print ("exp12 = %f" %exp12)
print ("exp34 = %f" %exp34)

exp12 = 52.326500
exp34 = 436.173500


In [35]:
convicted_ex = (0.1*568 + 0.117*409)/(568+409)
uncon_ex = (0.9*568 + 0.883*409)/(568+409)
Chi_sq = (((0.1*568)-568*convicted_ex)**2/(568*convicted_ex))+\
(((0.117*409)-409*convicted_ex)**2/(409*convicted_ex))+\
(((0.9*568)-568*uncon_ex)**2/(568*uncon_ex))+\
(((0.883*409)-(409*uncon_ex))**2/(409*uncon_ex))
Chi_sq

0.7184939175052877

### In this case, p = 0.05 and the DOF = 1, Chi_sq critical value we use here is 3.84. 0.718 < 3.84 so that we could not reject the NULL hypothesis. 